storagenode/piecestore/endpoint.go

```diff
@@ -765,3 +792,45 @@ func min(a, b int64) int64 {
 	}
 	return b
 }
+
+// speedEstimation monitors state of incoming traffic. It would signal slow-speed
+// client in non-congested traffic condition.
+type speedEstimation struct {
+	// grace indicates a certain period of time before the observator kicks in
+	grace time.Duration
+	// limit for flagging slow connections. Speed below this limit is considered to be slow.
+	limit memory.Size
+	// active indicates the duration of active connection.
+	active      time.Duration
```

Comment:
I don't think you need both `active` and `lastChecked`. `lastChecked` is only used to calculate `delta`, and `delta` is only used to increment `active`. Instead of either one, you could just have a `startedAt time.Time` member, fill it in on the first call to `EnsureLimit()`, and on every call after that you can calculate

    active = now.Sub(estimate.startedAt)

Reply:
That wouldn't allow to skip over time spent in congested state.

Reply:
Oh!! Ok, now I see the point. Perhaps the name could be something like `timeSpentNotCongested` to make it more clear?

Also, if we're only going to count deltas accumulated during "non-congested" time, we should do the same thing with the amount of data. This guess about congested-ness may not have anything to do with whether there is actually network congestion, or saturation, and there might still be a large amount of data transferred while isCongested() is true.

Imagine a node that is usually in the 'congested' state. I might connect to it from my uplink and start uploading at 1kB/s. slowly. 18 hours later, once the node happens to come out of 'congested' state for a while, I have managed to transfer 64MB. A few seconds later, this logic might say "wow thepaul's connection is going at 32MB/s, that's so fast" and leave it alone.

Reply:
`timeSpentNotCongested` might be slightly nicer choice. `nonCongestedTime` maybe?

The reason I wasn't ignoring the data transfer either, because the "congested" check isn't really atomic with regards to all other data we are gathering. So, by including the data transferred we err on the side of avoiding false-positives.

There of course are probably better ways to estimate those values. So, as long as uplink was able to make some decent progress, it's all fine I guess.

Reply:
`uncongestedTime` is probably better.